src/OpenRiaServices.DomainServices.Hosting/Framework/Services/MessageEncoders/PoxBinaryMessageEncodingBindingElement.cs

```diff
@@ -255,15 +256,235 @@ public override ArraySegment<byte> WriteMessage(Message message, int maxMessageS
+                /// Ensures _buffer is non null and has space to write more bytes
+                /// </summary>
+                private void EnsureBufferCapacity()
+                {
+                    // There is space left
+                    if (_bufferWritten < _buffer.Length)
+                        return;
+
+                    // Save current buffer in list before allocating a new buffer
+                    if (_bufferList == null)
+                        _bufferList = new System.Collections.Generic.List<byte[]>(capacity: 16);
+                    _bufferList.Add(_buffer);
+                    // Ensure we never return buffer twice in case TakeBuffer below throws
+                    _buffer = null;
+
+                    int nextSize = Math.Min(_position * 2, _maxSize);
+                    _buffer = _bufferManager.TakeBuffer(nextSize);
+                    _bufferWritten = 0;
+                }
+
+                protected override void Dispose(bool disposing)
+                {
+                    try
+                    {
+                        base.Dispose(disposing);
+                    }
+                    finally
+                    {
+                        if (_buffer != null)
+                        {
+                            _bufferManager.ReturnBuffer(_buffer);
+                            _buffer = null;
+                        }
+
+                        if (_bufferList != null)
+                        {
+                            foreach (var buffer in _bufferList)
+                                _bufferManager.ReturnBuffer(buffer);
+                            _bufferList = null;
+                        }
+                    }
+                }
+
+                /// <summary>
+                /// Copies bytes from <paramref name="src"/> to <paramref name="dest"/> using fastes 
+                /// copy based on process bitness (x86 / x64) tested on .Net Framework 4.8
+                /// </summary>
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                private static unsafe void FastCopy(byte[] src, int srcOffset, byte[] dest, int destOffset, int count)
+                {
+                    if (count == 0)
+                        return;
+
+                    if (Is64BitProcess && count <= 1024)
+                    {
+                        fixed (byte* s = &src[srcOffset], d = &dest[destOffset])
+                            Buffer.MemoryCopy(s, d, dest.Length - destOffset, count);
+                    }
+                    else
+                    {
+                        // For x86 it is significantly faster to do copying of int's and longs
+                        // or similar in managed code for smaller counts (below 100-200)
+                        // But we expect most copies to be larger since xml writer buffer around 500 bytes
+                        Buffer.BlockCopy(src, srcOffset, dest, destOffset, count);
```

Comment:
x86 copy speed 

Number of bytes | Buffer_BlockCopy | FastCopy_Long
-- | -- | --
4 | 25.402 | 6.404
40 | 30.176 | 22.202
200 | 65.845 | 54.938